<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source of the Rust file `src/process.rs`."><title>process.rs - source</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-1a91846b.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="hsc" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0 (6b00bc388 2025-06-23)" data-channel="1.88.0" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../../static.files/src-script-63605ae7.js"></script><script defer src="../../src-files.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc src"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="src-sidebar-title"><h2>Files</h2></div></nav><div class="sidebar-resizer"></div><main><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1><div class="sub-heading">hsc/</div>process.rs</h1><rustdoc-toolbar></rustdoc-toolbar></div><div class="example-wrap digits-3"><pre class="rust"><code><a href=#1 id=1 data-nosnippet>1</a><span class="kw">use </span><span class="kw">crate</span>::genotype::{MutationalBurden, Sfs};
<a href=#2 id=2 data-nosnippet>2</a><span class="kw">use </span><span class="kw">crate</span>::proliferation::{NeutralMutations, Proliferation};
<a href=#3 id=3 data-nosnippet>3</a><span class="kw">use </span><span class="kw">crate</span>::stemcell::{assign_background_mutations, StemCell};
<a href=#4 id=4 data-nosnippet>4</a><span class="kw">use </span><span class="kw">crate</span>::subclone::{save_variant_fraction, CloneId, Distributions, SubClones, Variants};
<a href=#5 id=5 data-nosnippet>5</a><span class="kw">use crate</span>::{MAX_SUBCLONES, TIME_AT_BIRTH};
<a href=#6 id=6 data-nosnippet>6</a><span class="kw">use </span>anyhow::Context;
<a href=#7 id=7 data-nosnippet>7</a><span class="kw">use </span>rand::Rng;
<a href=#8 id=8 data-nosnippet>8</a><span class="kw">use </span>rand_distr::weighted::WeightedIndex;
<a href=#9 id=9 data-nosnippet>9</a><span class="kw">use </span>rand_distr::Distribution;
<a href=#10 id=10 data-nosnippet>10</a><span class="kw">use </span>sosa::{AdvanceStep, CurrentState, NextReaction};
<a href=#11 id=11 data-nosnippet>11</a><span class="kw">use </span>std::collections::VecDeque;
<a href=#12 id=12 data-nosnippet>12</a><span class="kw">use </span>std::fs;
<a href=#13 id=13 data-nosnippet>13</a><span class="kw">use </span>std::num::NonZeroUsize;
<a href=#14 id=14 data-nosnippet>14</a><span class="kw">use </span>std::path::PathBuf;
<a href=#15 id=15 data-nosnippet>15</a>
<a href=#16 id=16 data-nosnippet>16</a><span class="attr">#[derive(Debug, Clone, Copy)]
<a href=#17 id=17 data-nosnippet>17</a></span><span class="kw">pub enum </span>SavingCells {
<a href=#18 id=18 data-nosnippet>18</a>    WholePopulation,
<a href=#19 id=19 data-nosnippet>19</a>    Subsampling {
<a href=#20 id=20 data-nosnippet>20</a>        sample_size: usize,
<a href=#21 id=21 data-nosnippet>21</a>        population_as_well: bool,
<a href=#22 id=22 data-nosnippet>22</a>    },
<a href=#23 id=23 data-nosnippet>23</a>}
<a href=#24 id=24 data-nosnippet>24</a>
<a href=#25 id=25 data-nosnippet>25</a><span class="attr">#[derive(Debug, Clone)]
<a href=#26 id=26 data-nosnippet>26</a></span><span class="kw">pub struct </span>Snapshot {
<a href=#27 id=27 data-nosnippet>27</a>    <span class="doccomment">/// The number of cells to subsample
<a href=#28 id=28 data-nosnippet>28</a>    </span><span class="kw">pub </span>cells2sample: usize,
<a href=#29 id=29 data-nosnippet>29</a>    <span class="doccomment">/// The time at which we subsample
<a href=#30 id=30 data-nosnippet>30</a>    </span><span class="kw">pub </span>time: f32,
<a href=#31 id=31 data-nosnippet>31</a>}
<a href=#32 id=32 data-nosnippet>32</a>
<a href=#33 id=33 data-nosnippet>33</a><span class="attr">#[derive(Debug, Clone)]
<a href=#34 id=34 data-nosnippet>34</a></span><span class="kw">pub struct </span>SavingOptions {
<a href=#35 id=35 data-nosnippet>35</a>    <span class="kw">pub </span>filename: PathBuf,
<a href=#36 id=36 data-nosnippet>36</a>    <span class="kw">pub </span>save_sfs_only: bool,
<a href=#37 id=37 data-nosnippet>37</a>    <span class="kw">pub </span>save_population: bool,
<a href=#38 id=38 data-nosnippet>38</a>}
<a href=#39 id=39 data-nosnippet>39</a>
<a href=#40 id=40 data-nosnippet>40</a><span class="attr">#[derive(Debug, Clone, Default)]
<a href=#41 id=41 data-nosnippet>41</a></span><span class="kw">pub struct </span>ProcessOptions {
<a href=#42 id=42 data-nosnippet>42</a>    <span class="kw">pub </span>path: PathBuf,
<a href=#43 id=43 data-nosnippet>43</a>    <span class="kw">pub </span>snapshots: VecDeque&lt;Snapshot&gt;,
<a href=#44 id=44 data-nosnippet>44</a>}
<a href=#45 id=45 data-nosnippet>45</a>
<a href=#46 id=46 data-nosnippet>46</a><span class="attr">#[derive(Hash, PartialEq, Eq)]
<a href=#47 id=47 data-nosnippet>47</a></span><span class="kw">pub enum </span>Stats2Save {
<a href=#48 id=48 data-nosnippet>48</a>    Burden,
<a href=#49 id=49 data-nosnippet>49</a>    Sfs,
<a href=#50 id=50 data-nosnippet>50</a>    VariantFraction,
<a href=#51 id=51 data-nosnippet>51</a>}
<a href=#52 id=52 data-nosnippet>52</a>
<a href=#53 id=53 data-nosnippet>53</a><span class="attr">#[derive(Debug, Clone)]
<a href=#54 id=54 data-nosnippet>54</a></span><span class="kw">pub struct </span>CellDivisionProbabilities {
<a href=#55 id=55 data-nosnippet>55</a>    <span class="doccomment">/// Rate of neutral mutations per cell-division
<a href=#56 id=56 data-nosnippet>56</a>    </span><span class="kw">pub </span>lambda_poisson: f32,
<a href=#57 id=57 data-nosnippet>57</a>    <span class="doccomment">/// Probability of getting a fit mutant upon cell division
<a href=#58 id=58 data-nosnippet>58</a>    </span><span class="kw">pub </span>p: f64,
<a href=#59 id=59 data-nosnippet>59</a>}
<a href=#60 id=60 data-nosnippet>60</a>
<a href=#61 id=61 data-nosnippet>61</a><span class="attr">#[derive(Clone, Debug)]
<a href=#62 id=62 data-nosnippet>62</a></span><span class="kw">pub enum </span>ExponentialPhase {
<a href=#63 id=63 data-nosnippet>63</a>    <span class="doccomment">/// The embryonic phase up to birth
<a href=#64 id=64 data-nosnippet>64</a>    </span>Development,
<a href=#65 id=65 data-nosnippet>65</a>    <span class="doccomment">/// A regrowth exponential phase upon subsampling or treatment.
<a href=#66 id=66 data-nosnippet>66</a>    </span>Regrowth,
<a href=#67 id=67 data-nosnippet>67</a>}
<a href=#68 id=68 data-nosnippet>68</a>
<a href=#69 id=69 data-nosnippet>69</a><span class="kw">pub fn </span>switch_to_moran(
<a href=#70 id=70 data-nosnippet>70</a>    <span class="kw-2">mut </span>exponential: Exponential,
<a href=#71 id=71 data-nosnippet>71</a>    process_options: ProcessOptions,
<a href=#72 id=72 data-nosnippet>72</a>    distributions: Distributions,
<a href=#73 id=73 data-nosnippet>73</a>    filename: PathBuf,
<a href=#74 id=74 data-nosnippet>74</a>    save_sfs_only: bool,
<a href=#75 id=75 data-nosnippet>75</a>    save_population: bool,
<a href=#76 id=76 data-nosnippet>76</a>    rng: <span class="kw-2">&amp;mut </span><span class="kw">impl </span>Rng,
<a href=#77 id=77 data-nosnippet>77</a>) -&gt; Moran {
<a href=#78 id=78 data-nosnippet>78</a>    <span class="doccomment">//! End the exponential growing phase and switch to a fixed-size
<a href=#79 id=79 data-nosnippet>79</a>    //! population phase.
<a href=#80 id=80 data-nosnippet>80</a>    //!
<a href=#81 id=81 data-nosnippet>81</a>    //! There is a delay between birth and the end of the exponential phase,
<a href=#82 id=82 data-nosnippet>82</a>    //! since stem cells stop dividing exponentially before birth.
<a href=#83 id=83 data-nosnippet>83</a>    //! We add background mutations in this interval of time between the
<a href=#84 id=84 data-nosnippet>84</a>    //! end of the exponentially growing phase and the Moran process.
<a href=#85 id=85 data-nosnippet>85</a>    </span><span class="kw">if </span>exponential.verbosity &gt; <span class="number">0 </span>{
<a href=#86 id=86 data-nosnippet>86</a>        <span class="macro">println!</span>(<span class="string">"switching to Moran at time {}"</span>, exponential.time);
<a href=#87 id=87 data-nosnippet>87</a>    }
<a href=#88 id=88 data-nosnippet>88</a>    <span class="kw">let </span>(current_time, time_to_restart) = <span class="kw">match </span>exponential.phase {
<a href=#89 id=89 data-nosnippet>89</a>        ExponentialPhase::Development =&gt; (TIME_AT_BIRTH, <span class="number">0f32</span>),
<a href=#90 id=90 data-nosnippet>90</a>        ExponentialPhase::Regrowth =&gt; (exponential.time, exponential.time),
<a href=#91 id=91 data-nosnippet>91</a>    };
<a href=#92 id=92 data-nosnippet>92</a>    <span class="kw">if let </span>NeutralMutations::UponDivisionAndBackground = exponential.proliferation.neutral_mutation
<a href=#93 id=93 data-nosnippet>93</a>    {
<a href=#94 id=94 data-nosnippet>94</a>        <span class="comment">// this is important: we update all background mutations at this
<a href=#95 id=95 data-nosnippet>95</a>        // time such that all cells are all on the same page.
<a href=#96 id=96 data-nosnippet>96</a>        // Since background mutations are implemented at each division, and
<a href=#97 id=97 data-nosnippet>97</a>        // cells do not proliferate at the same rate, we need to correct and
<a href=#98 id=98 data-nosnippet>98</a>        // update the background mutations at the timepoint corresponding
<a href=#99 id=99 data-nosnippet>99</a>        // to sampling step, i.e. before saving.
<a href=#100 id=100 data-nosnippet>100</a>        //
<a href=#101 id=101 data-nosnippet>101</a>        // Moreover there is delay between the end of the exp. growing phase
<a href=#102 id=102 data-nosnippet>102</a>        // and birth, hence we use `TIME_AT_BIRTH`. But this is true only for
<a href=#103 id=103 data-nosnippet>103</a>        // the Development phase.
<a href=#104 id=104 data-nosnippet>104</a>        </span><span class="kw">if </span>exponential.verbosity &gt; <span class="number">0 </span>{
<a href=#105 id=105 data-nosnippet>105</a>            <span class="macro">println!</span>(<span class="string">"updating the neutral background mutations for all cells"</span>);
<a href=#106 id=106 data-nosnippet>106</a>        }
<a href=#107 id=107 data-nosnippet>107</a>        <span class="comment">// assign missing mutations
<a href=#108 id=108 data-nosnippet>108</a>        </span><span class="kw">for </span>stem_cell <span class="kw">in </span>exponential.subclones.get_mut_cells() {
<a href=#109 id=109 data-nosnippet>109</a>            <span class="kw">if </span>stem_cell.get_last_division_time() &lt; <span class="kw-2">&amp;</span>current_time {
<a href=#110 id=110 data-nosnippet>110</a>                assign_background_mutations(
<a href=#111 id=111 data-nosnippet>111</a>                    stem_cell,
<a href=#112 id=112 data-nosnippet>112</a>                    current_time,
<a href=#113 id=113 data-nosnippet>113</a>                    <span class="kw-2">&amp;</span>exponential.distributions.neutral_poisson,
<a href=#114 id=114 data-nosnippet>114</a>                    rng,
<a href=#115 id=115 data-nosnippet>115</a>                    exponential.verbosity,
<a href=#116 id=116 data-nosnippet>116</a>                );
<a href=#117 id=117 data-nosnippet>117</a>            }
<a href=#118 id=118 data-nosnippet>118</a>            <span class="comment">// restarting the time within cells
<a href=#119 id=119 data-nosnippet>119</a>            </span>stem_cell.set_last_division_time(time_to_restart).unwrap();
<a href=#120 id=120 data-nosnippet>120</a>        }
<a href=#121 id=121 data-nosnippet>121</a>    }
<a href=#122 id=122 data-nosnippet>122</a>    Moran {
<a href=#123 id=123 data-nosnippet>123</a>        subclones: exponential.subclones,
<a href=#124 id=124 data-nosnippet>124</a>        counter_divisions: exponential.counter_divisions,
<a href=#125 id=125 data-nosnippet>125</a>        <span class="comment">// restart the time within the process
<a href=#126 id=126 data-nosnippet>126</a>        </span>time: time_to_restart,
<a href=#127 id=127 data-nosnippet>127</a>        snapshots: process_options.snapshots,
<a href=#128 id=128 data-nosnippet>128</a>        path2dir: process_options.path,
<a href=#129 id=129 data-nosnippet>129</a>        verbosity: exponential.verbosity,
<a href=#130 id=130 data-nosnippet>130</a>        filename,
<a href=#131 id=131 data-nosnippet>131</a>        distributions,
<a href=#132 id=132 data-nosnippet>132</a>        save_sfs_only,
<a href=#133 id=133 data-nosnippet>133</a>        save_population,
<a href=#134 id=134 data-nosnippet>134</a>        proliferation: exponential.proliferation,
<a href=#135 id=135 data-nosnippet>135</a>    }
<a href=#136 id=136 data-nosnippet>136</a>}
<a href=#137 id=137 data-nosnippet>137</a>
<a href=#138 id=138 data-nosnippet>138</a><span class="doccomment">/// Exponential growing process
<a href=#139 id=139 data-nosnippet>139</a></span><span class="attr">#[derive(Clone, Debug)]
<a href=#140 id=140 data-nosnippet>140</a></span><span class="kw">pub struct </span>Exponential {
<a href=#141 id=141 data-nosnippet>141</a>    <span class="doccomment">/// A collection of clones having a proliferative advantage.
<a href=#142 id=142 data-nosnippet>142</a>    </span><span class="kw">pub </span>subclones: SubClones,
<a href=#143 id=143 data-nosnippet>143</a>    <span class="doccomment">/// The counter for the number of proliferative events.
<a href=#144 id=144 data-nosnippet>144</a>    </span><span class="kw">pub </span>counter_divisions: usize,
<a href=#145 id=145 data-nosnippet>145</a>    <span class="kw">pub </span>verbosity: u8,
<a href=#146 id=146 data-nosnippet>146</a>    <span class="kw">pub </span>distributions: Distributions,
<a href=#147 id=147 data-nosnippet>147</a>    <span class="kw">pub </span>proliferation: Proliferation,
<a href=#148 id=148 data-nosnippet>148</a>    <span class="kw">pub </span>time: f32,
<a href=#149 id=149 data-nosnippet>149</a>    <span class="kw">pub </span>phase: ExponentialPhase,
<a href=#150 id=150 data-nosnippet>150</a>}
<a href=#151 id=151 data-nosnippet>151</a>
<a href=#152 id=152 data-nosnippet>152</a><span class="kw">impl </span>Exponential {
<a href=#153 id=153 data-nosnippet>153</a>    <span class="kw">pub fn </span>new(
<a href=#154 id=154 data-nosnippet>154</a>        initial_subclones: SubClones,
<a href=#155 id=155 data-nosnippet>155</a>        distributions: Distributions,
<a href=#156 id=156 data-nosnippet>156</a>        proliferation: Proliferation,
<a href=#157 id=157 data-nosnippet>157</a>        time: f32,
<a href=#158 id=158 data-nosnippet>158</a>        phase: ExponentialPhase,
<a href=#159 id=159 data-nosnippet>159</a>        verbosity: u8,
<a href=#160 id=160 data-nosnippet>160</a>    ) -&gt; Exponential {
<a href=#161 id=161 data-nosnippet>161</a>        <span class="kw">let </span>hsc = Exponential {
<a href=#162 id=162 data-nosnippet>162</a>            subclones: initial_subclones,
<a href=#163 id=163 data-nosnippet>163</a>            distributions,
<a href=#164 id=164 data-nosnippet>164</a>            counter_divisions: <span class="number">0</span>,
<a href=#165 id=165 data-nosnippet>165</a>            proliferation,
<a href=#166 id=166 data-nosnippet>166</a>            verbosity,
<a href=#167 id=167 data-nosnippet>167</a>            phase,
<a href=#168 id=168 data-nosnippet>168</a>            time,
<a href=#169 id=169 data-nosnippet>169</a>        };
<a href=#170 id=170 data-nosnippet>170</a>        <span class="kw">if </span>verbosity &gt; <span class="number">1 </span>{
<a href=#171 id=171 data-nosnippet>171</a>            <span class="macro">println!</span>(<span class="string">"process created: {hsc:#?}"</span>);
<a href=#172 id=172 data-nosnippet>172</a>        }
<a href=#173 id=173 data-nosnippet>173</a>        hsc
<a href=#174 id=174 data-nosnippet>174</a>    }
<a href=#175 id=175 data-nosnippet>175</a>}
<a href=#176 id=176 data-nosnippet>176</a>
<a href=#177 id=177 data-nosnippet>177</a><span class="kw">impl </span>From&lt;Moran&gt; <span class="kw">for </span>Exponential {
<a href=#178 id=178 data-nosnippet>178</a>    <span class="kw">fn </span>from(value: Moran) -&gt; <span class="self">Self </span>{
<a href=#179 id=179 data-nosnippet>179</a>        <span class="doccomment">//! Construct an exponential-growing process in [`ExponentialPhase::Regrowth`] mode.
<a href=#180 id=180 data-nosnippet>180</a>        //!
<a href=#181 id=181 data-nosnippet>181</a>        //! This is useful when we undersample the Moran process to a smaller
<a href=#182 id=182 data-nosnippet>182</a>        //! population size (for example upon treatment, the number of cells is
<a href=#183 id=183 data-nosnippet>183</a>        //! reduced) and we want to grow back the population to its size before
<a href=#184 id=184 data-nosnippet>184</a>        //! the undersampling.
<a href=#185 id=185 data-nosnippet>185</a>        </span>Exponential::new(
<a href=#186 id=186 data-nosnippet>186</a>            value.subclones,
<a href=#187 id=187 data-nosnippet>187</a>            value.distributions,
<a href=#188 id=188 data-nosnippet>188</a>            value.proliferation,
<a href=#189 id=189 data-nosnippet>189</a>            value.time,
<a href=#190 id=190 data-nosnippet>190</a>            ExponentialPhase::Regrowth,
<a href=#191 id=191 data-nosnippet>191</a>            value.verbosity,
<a href=#192 id=192 data-nosnippet>192</a>        )
<a href=#193 id=193 data-nosnippet>193</a>    }
<a href=#194 id=194 data-nosnippet>194</a>}
<a href=#195 id=195 data-nosnippet>195</a>
<a href=#196 id=196 data-nosnippet>196</a><span class="kw">impl </span>AdvanceStep&lt;MAX_SUBCLONES&gt; <span class="kw">for </span>Exponential {
<a href=#197 id=197 data-nosnippet>197</a>    <span class="doccomment">/// The id of a subclone [`CloneId`].
<a href=#198 id=198 data-nosnippet>198</a>    </span><span class="kw">type </span>Reaction = CloneId;
<a href=#199 id=199 data-nosnippet>199</a>
<a href=#200 id=200 data-nosnippet>200</a>    <span class="kw">fn </span>advance_step(<span class="kw-2">&amp;mut </span><span class="self">self</span>, reaction: NextReaction&lt;<span class="self">Self</span>::Reaction&gt;, rng: <span class="kw-2">&amp;mut </span><span class="kw">impl </span>Rng) {
<a href=#201 id=201 data-nosnippet>201</a>        <span class="doccomment">//! Update the process by simulating the next proliferative event
<a href=#202 id=202 data-nosnippet>202</a>        //! according to the next `reaction` determined by the [Gillespie
<a href=#203 id=203 data-nosnippet>203</a>        //! algorithm](`sosa`).
<a href=#204 id=204 data-nosnippet>204</a>        //!
<a href=#205 id=205 data-nosnippet>205</a>        //! The proliferation step is implemented as following:
<a href=#206 id=206 data-nosnippet>206</a>        //!
<a href=#207 id=207 data-nosnippet>207</a>        //! 1. select the cell that will proliferate next from the clone
<a href=#208 id=208 data-nosnippet>208</a>        //!    with id `reaction` determined by the Gillespie algorithm
<a href=#209 id=209 data-nosnippet>209</a>        //!
<a href=#210 id=210 data-nosnippet>210</a>        //! 2. for the proliferating cell and its daughter cell:
<a href=#211 id=211 data-nosnippet>211</a>        //!     * mutate genome by storing a neutral number of mutations
<a href=#212 id=212 data-nosnippet>212</a>        //!
<a href=#213 id=213 data-nosnippet>213</a>        //!     * assign to new subclone with a probability determined by the
<a href=#214 id=214 data-nosnippet>214</a>        //!       rate of mutations conferring a proliferative advantage
<a href=#215 id=215 data-nosnippet>215</a>        </span><span class="comment">// The Gillespie sampler samples the clone that will proliferate next,
<a href=#216 id=216 data-nosnippet>216</a>        // that is the clone with id `reaction.event`.
<a href=#217 id=217 data-nosnippet>217</a>        // Pick random proliferating cells from this clone. **Note that this
<a href=#218 id=218 data-nosnippet>218</a>        // removes the cells from the clone with id `reaction.event`.**
<a href=#219 id=219 data-nosnippet>219</a>        </span><span class="self">self</span>.counter_divisions += <span class="number">1</span>;
<a href=#220 id=220 data-nosnippet>220</a>        <span class="self">self</span>.time += reaction.time;
<a href=#221 id=221 data-nosnippet>221</a>        <span class="self">self</span>.proliferation.proliferate(
<a href=#222 id=222 data-nosnippet>222</a>            <span class="kw-2">&amp;mut </span><span class="self">self</span>.subclones,
<a href=#223 id=223 data-nosnippet>223</a>            <span class="self">self</span>.time,
<a href=#224 id=224 data-nosnippet>224</a>            reaction.event,
<a href=#225 id=225 data-nosnippet>225</a>            <span class="kw-2">&amp;</span><span class="self">self</span>.distributions,
<a href=#226 id=226 data-nosnippet>226</a>            rng,
<a href=#227 id=227 data-nosnippet>227</a>            <span class="self">self</span>.verbosity,
<a href=#228 id=228 data-nosnippet>228</a>        );
<a href=#229 id=229 data-nosnippet>229</a>    }
<a href=#230 id=230 data-nosnippet>230</a>
<a href=#231 id=231 data-nosnippet>231</a>    <span class="kw">fn </span>update_state(<span class="kw-2">&amp;</span><span class="self">self</span>, state: <span class="kw-2">&amp;mut </span>CurrentState&lt;MAX_SUBCLONES&gt;) {
<a href=#232 id=232 data-nosnippet>232</a>        state.population = Variants::variant_counts(<span class="kw-2">&amp;</span><span class="self">self</span>.subclones);
<a href=#233 id=233 data-nosnippet>233</a>    }
<a href=#234 id=234 data-nosnippet>234</a>}
<a href=#235 id=235 data-nosnippet>235</a>
<a href=#236 id=236 data-nosnippet>236</a><span class="doccomment">/// The Moran process saves the state of the agents and simulates new
<a href=#237 id=237 data-nosnippet>237</a>/// proliferative events at each timestep according to the Gillespie algorithm,
<a href=#238 id=238 data-nosnippet>238</a>/// see [`Moran::advance_step`].
<a href=#239 id=239 data-nosnippet>239</a></span><span class="attr">#[derive(Debug, Clone)]
<a href=#240 id=240 data-nosnippet>240</a></span><span class="kw">pub struct </span>Moran {
<a href=#241 id=241 data-nosnippet>241</a>    <span class="doccomment">/// A collection of clones having a proliferative advantage.
<a href=#242 id=242 data-nosnippet>242</a>    </span><span class="kw">pub </span>subclones: SubClones,
<a href=#243 id=243 data-nosnippet>243</a>    <span class="doccomment">/// The counter for the number of proliferative events.
<a href=#244 id=244 data-nosnippet>244</a>    </span><span class="kw">pub </span>counter_divisions: usize,
<a href=#245 id=245 data-nosnippet>245</a>    <span class="kw">pub </span>time: f32,
<a href=#246 id=246 data-nosnippet>246</a>    <span class="kw">pub </span>path2dir: PathBuf,
<a href=#247 id=247 data-nosnippet>247</a>    <span class="kw">pub </span>verbosity: u8,
<a href=#248 id=248 data-nosnippet>248</a>    <span class="kw">pub </span>distributions: Distributions,
<a href=#249 id=249 data-nosnippet>249</a>    <span class="kw">pub </span>snapshots: VecDeque&lt;Snapshot&gt;,
<a href=#250 id=250 data-nosnippet>250</a>    <span class="kw">pub </span>filename: PathBuf,
<a href=#251 id=251 data-nosnippet>251</a>    <span class="kw">pub </span>save_sfs_only: bool,
<a href=#252 id=252 data-nosnippet>252</a>    <span class="kw">pub </span>save_population: bool,
<a href=#253 id=253 data-nosnippet>253</a>    <span class="kw">pub </span>proliferation: Proliferation,
<a href=#254 id=254 data-nosnippet>254</a>}
<a href=#255 id=255 data-nosnippet>255</a>
<a href=#256 id=256 data-nosnippet>256</a><span class="kw">impl </span>Default <span class="kw">for </span>Moran {
<a href=#257 id=257 data-nosnippet>257</a>    <span class="kw">fn </span>default() -&gt; <span class="self">Self </span>{
<a href=#258 id=258 data-nosnippet>258</a>        <span class="kw">let </span>process_options = ProcessOptions {
<a href=#259 id=259 data-nosnippet>259</a>            path: PathBuf::from(<span class="string">"./output"</span>),
<a href=#260 id=260 data-nosnippet>260</a>            snapshots: VecDeque::default(),
<a href=#261 id=261 data-nosnippet>261</a>        };
<a href=#262 id=262 data-nosnippet>262</a>
<a href=#263 id=263 data-nosnippet>263</a>        Moran::new(
<a href=#264 id=264 data-nosnippet>264</a>            process_options,
<a href=#265 id=265 data-nosnippet>265</a>            SubClones::default(),
<a href=#266 id=266 data-nosnippet>266</a>            <span class="number">0.</span>,
<a href=#267 id=267 data-nosnippet>267</a>            SavingOptions {
<a href=#268 id=268 data-nosnippet>268</a>                filename: PathBuf::default(),
<a href=#269 id=269 data-nosnippet>269</a>                save_sfs_only: <span class="bool-val">false</span>,
<a href=#270 id=270 data-nosnippet>270</a>                save_population: <span class="bool-val">true</span>,
<a href=#271 id=271 data-nosnippet>271</a>            },
<a href=#272 id=272 data-nosnippet>272</a>            Distributions::default(),
<a href=#273 id=273 data-nosnippet>273</a>            Proliferation::default(),
<a href=#274 id=274 data-nosnippet>274</a>            <span class="number">1</span>,
<a href=#275 id=275 data-nosnippet>275</a>        )
<a href=#276 id=276 data-nosnippet>276</a>    }
<a href=#277 id=277 data-nosnippet>277</a>}
<a href=#278 id=278 data-nosnippet>278</a>
<a href=#279 id=279 data-nosnippet>279</a><span class="kw">impl </span>Moran {
<a href=#280 id=280 data-nosnippet>280</a>    <span class="doccomment">/// A Moran process with wild-type subclone with neutral fitness, to which
<a href=#281 id=281 data-nosnippet>281</a>    /// all cells will be assigned.
<a href=#282 id=282 data-nosnippet>282</a>    </span><span class="kw">pub fn </span>new(
<a href=#283 id=283 data-nosnippet>283</a>        process_options: ProcessOptions,
<a href=#284 id=284 data-nosnippet>284</a>        initial_subclones: SubClones,
<a href=#285 id=285 data-nosnippet>285</a>        time: f32,
<a href=#286 id=286 data-nosnippet>286</a>        saving_options: SavingOptions,
<a href=#287 id=287 data-nosnippet>287</a>        distributions: Distributions,
<a href=#288 id=288 data-nosnippet>288</a>        proliferation: Proliferation,
<a href=#289 id=289 data-nosnippet>289</a>        verbosity: u8,
<a href=#290 id=290 data-nosnippet>290</a>    ) -&gt; Moran {
<a href=#291 id=291 data-nosnippet>291</a>        <span class="kw">let </span>hsc = Moran {
<a href=#292 id=292 data-nosnippet>292</a>            subclones: initial_subclones,
<a href=#293 id=293 data-nosnippet>293</a>            distributions,
<a href=#294 id=294 data-nosnippet>294</a>            counter_divisions: <span class="number">0</span>,
<a href=#295 id=295 data-nosnippet>295</a>            path2dir: process_options.path,
<a href=#296 id=296 data-nosnippet>296</a>            time,
<a href=#297 id=297 data-nosnippet>297</a>            snapshots: process_options.snapshots,
<a href=#298 id=298 data-nosnippet>298</a>            filename: saving_options.filename,
<a href=#299 id=299 data-nosnippet>299</a>            verbosity,
<a href=#300 id=300 data-nosnippet>300</a>            save_sfs_only: saving_options.save_sfs_only,
<a href=#301 id=301 data-nosnippet>301</a>            save_population: saving_options.save_population,
<a href=#302 id=302 data-nosnippet>302</a>            proliferation,
<a href=#303 id=303 data-nosnippet>303</a>        };
<a href=#304 id=304 data-nosnippet>304</a>        <span class="kw">if </span>verbosity &gt; <span class="number">1 </span>{
<a href=#305 id=305 data-nosnippet>305</a>            <span class="macro">println!</span>(<span class="string">"process created: {hsc:#?}"</span>);
<a href=#306 id=306 data-nosnippet>306</a>        }
<a href=#307 id=307 data-nosnippet>307</a>        hsc
<a href=#308 id=308 data-nosnippet>308</a>    }
<a href=#309 id=309 data-nosnippet>309</a>
<a href=#310 id=310 data-nosnippet>310</a>    <span class="kw">fn </span>keep_const_population_upon_symmetric_division(<span class="kw-2">&amp;mut </span><span class="self">self</span>, rng: <span class="kw-2">&amp;mut </span><span class="kw">impl </span>Rng) {
<a href=#311 id=311 data-nosnippet>311</a>        <span class="doccomment">//! If an symmetric division is performed, need to remove a random cell
<a href=#312 id=312 data-nosnippet>312</a>        //! from the population.
<a href=#313 id=313 data-nosnippet>313</a>        //! The cell having just proliferated can be removed as well, any cell
<a href=#314 id=314 data-nosnippet>314</a>        //! can be removed.
<a href=#315 id=315 data-nosnippet>315</a>        //!
<a href=#316 id=316 data-nosnippet>316</a>        //! We proceed as following:
<a href=#317 id=317 data-nosnippet>317</a>        //!     1. check if there is only one clone in the population, then
<a href=#318 id=318 data-nosnippet>318</a>        //!     remove a cell from this clone
<a href=#319 id=319 data-nosnippet>319</a>        //!     2. else, compute the variant counts
<a href=#320 id=320 data-nosnippet>320</a>        //!     3. and sample from any clone based on the weights defined by
<a href=#321 id=321 data-nosnippet>321</a>        //!     the variant counts
<a href=#322 id=322 data-nosnippet>322</a>        </span><span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">2 </span>{
<a href=#323 id=323 data-nosnippet>323</a>            <span class="macro">println!</span>(<span class="string">"keeping the cell population constant"</span>);
<a href=#324 id=324 data-nosnippet>324</a>        }
<a href=#325 id=325 data-nosnippet>325</a>        <span class="comment">// remove a cell from a random subclone based on the frequencies of
<a href=#326 id=326 data-nosnippet>326</a>        // the clones at the current state
<a href=#327 id=327 data-nosnippet>327</a>        </span><span class="kw">let </span>variants = Variants::variant_counts(<span class="kw-2">&amp;</span><span class="self">self</span>.subclones);
<a href=#328 id=328 data-nosnippet>328</a>        <span class="kw">let </span>id2remove = WeightedIndex::new(variants).unwrap().sample(rng);
<a href=#329 id=329 data-nosnippet>329</a>        <span class="kw">let _ </span>= <span class="self">self
<a href=#330 id=330 data-nosnippet>330</a>            </span>.subclones
<a href=#331 id=331 data-nosnippet>331</a>            .get_mut_clone_unchecked(id2remove)
<a href=#332 id=332 data-nosnippet>332</a>            .random_cell(rng)
<a href=#333 id=333 data-nosnippet>333</a>            .with_context(|| <span class="string">"found empty subclone"</span>)
<a href=#334 id=334 data-nosnippet>334</a>            .unwrap();
<a href=#335 id=335 data-nosnippet>335</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">2 </span>{
<a href=#336 id=336 data-nosnippet>336</a>            <span class="macro">println!</span>(<span class="string">"removing one cell from clone {id2remove}"</span>);
<a href=#337 id=337 data-nosnippet>337</a>        }
<a href=#338 id=338 data-nosnippet>338</a>    }
<a href=#339 id=339 data-nosnippet>339</a>
<a href=#340 id=340 data-nosnippet>340</a>    <span class="kw">pub</span>(<span class="kw">crate</span>) <span class="kw">fn </span>make_path(
<a href=#341 id=341 data-nosnippet>341</a>        <span class="kw-2">&amp;</span><span class="self">self</span>,
<a href=#342 id=342 data-nosnippet>342</a>        tosave: Stats2Save,
<a href=#343 id=343 data-nosnippet>343</a>        cells: usize,
<a href=#344 id=344 data-nosnippet>344</a>        time: f32,
<a href=#345 id=345 data-nosnippet>345</a>    ) -&gt; anyhow::Result&lt;PathBuf&gt; {
<a href=#346 id=346 data-nosnippet>346</a>        <span class="kw">let </span>path2dir = <span class="self">self</span>.path2dir.join(<span class="macro">format!</span>(<span class="string">"{cells}cells"</span>));
<a href=#347 id=347 data-nosnippet>347</a>        <span class="kw">let </span>path2file = <span class="kw">match </span>tosave {
<a href=#348 id=348 data-nosnippet>348</a>            Stats2Save::VariantFraction =&gt; path2dir.join(<span class="string">"variant_fraction"</span>),
<a href=#349 id=349 data-nosnippet>349</a>            Stats2Save::Burden =&gt; path2dir.join(<span class="string">"burden"</span>),
<a href=#350 id=350 data-nosnippet>350</a>            Stats2Save::Sfs =&gt; path2dir.join(<span class="string">"sfs"</span>),
<a href=#351 id=351 data-nosnippet>351</a>        };
<a href=#352 id=352 data-nosnippet>352</a>        <span class="kw">let </span><span class="kw-2">mut </span>timepoint = <span class="macro">format!</span>(<span class="string">"{time:.1}"</span>).replace(<span class="string">'.'</span>, <span class="string">"dot"</span>);
<a href=#353 id=353 data-nosnippet>353</a>        timepoint.push_str(<span class="string">"years"</span>);
<a href=#354 id=354 data-nosnippet>354</a>        <span class="kw">let </span>path2file = path2file.join(timepoint);
<a href=#355 id=355 data-nosnippet>355</a>        fs::create_dir_all(<span class="kw-2">&amp;</span>path2file).with_context(|| <span class="string">"Cannot create dir"</span>)<span class="question-mark">?</span>;
<a href=#356 id=356 data-nosnippet>356</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">1 </span>{
<a href=#357 id=357 data-nosnippet>357</a>            <span class="macro">println!</span>(<span class="string">"creating dirs {path2file:#?}"</span>);
<a href=#358 id=358 data-nosnippet>358</a>        }
<a href=#359 id=359 data-nosnippet>359</a>        <span class="prelude-val">Ok</span>(path2file.join(<span class="self">self</span>.filename.clone()))
<a href=#360 id=360 data-nosnippet>360</a>    }
<a href=#361 id=361 data-nosnippet>361</a>
<a href=#362 id=362 data-nosnippet>362</a>    <span class="kw">fn </span>save_it(
<a href=#363 id=363 data-nosnippet>363</a>        <span class="kw-2">&amp;</span><span class="self">self</span>,
<a href=#364 id=364 data-nosnippet>364</a>        time: f32,
<a href=#365 id=365 data-nosnippet>365</a>        cells_with_idx: Vec&lt;(<span class="kw-2">&amp;</span>StemCell, usize)&gt;,
<a href=#366 id=366 data-nosnippet>366</a>        save_sfs_only: bool,
<a href=#367 id=367 data-nosnippet>367</a>    ) -&gt; anyhow::Result&lt;()&gt; {
<a href=#368 id=368 data-nosnippet>368</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#369 id=369 data-nosnippet>369</a>            <span class="macro">println!</span>(<span class="string">"saving data at time {time}"</span>);
<a href=#370 id=370 data-nosnippet>370</a>        }
<a href=#371 id=371 data-nosnippet>371</a>        <span class="kw">let </span>cells: Vec&lt;<span class="kw-2">&amp;</span>StemCell&gt; = cells_with_idx.iter().map(|ele| ele.<span class="number">0</span>).collect();
<a href=#372 id=372 data-nosnippet>372</a>        <span class="kw">let </span>nb_cells = cells.len();
<a href=#373 id=373 data-nosnippet>373</a>
<a href=#374 id=374 data-nosnippet>374</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#375 id=375 data-nosnippet>375</a>            <span class="macro">println!</span>(<span class="string">"saving {nb_cells} cells"</span>);
<a href=#376 id=376 data-nosnippet>376</a>        }
<a href=#377 id=377 data-nosnippet>377</a>
<a href=#378 id=378 data-nosnippet>378</a>        Sfs::from_cells(<span class="kw-2">&amp;</span>cells, <span class="self">self</span>.verbosity)
<a href=#379 id=379 data-nosnippet>379</a>            .unwrap_or_else(|<span class="kw">_</span>| <span class="macro">panic!</span>(<span class="string">"cannot create SFS for timepoint at time {time}"</span>))
<a href=#380 id=380 data-nosnippet>380</a>            .save(
<a href=#381 id=381 data-nosnippet>381</a>                <span class="kw-2">&amp;</span><span class="self">self</span>.make_path(Stats2Save::Sfs, nb_cells, time)<span class="question-mark">?</span>,
<a href=#382 id=382 data-nosnippet>382</a>                <span class="self">self</span>.verbosity,
<a href=#383 id=383 data-nosnippet>383</a>            )<span class="question-mark">?</span>;
<a href=#384 id=384 data-nosnippet>384</a>
<a href=#385 id=385 data-nosnippet>385</a>        <span class="kw">if </span>!save_sfs_only {
<a href=#386 id=386 data-nosnippet>386</a>            MutationalBurden::from_cells(<span class="kw-2">&amp;</span>cells, <span class="self">self</span>.verbosity)
<a href=#387 id=387 data-nosnippet>387</a>                .unwrap_or_else(|<span class="kw">_</span>| <span class="macro">panic!</span>(<span class="string">"cannot create burden for the timepoint at time {time}"</span>))
<a href=#388 id=388 data-nosnippet>388</a>                .save(
<a href=#389 id=389 data-nosnippet>389</a>                    <span class="kw-2">&amp;</span><span class="self">self</span>.make_path(Stats2Save::Burden, nb_cells, time)<span class="question-mark">?</span>,
<a href=#390 id=390 data-nosnippet>390</a>                    <span class="self">self</span>.verbosity,
<a href=#391 id=391 data-nosnippet>391</a>                )<span class="question-mark">?</span>;
<a href=#392 id=392 data-nosnippet>392</a>            save_variant_fraction(
<a href=#393 id=393 data-nosnippet>393</a>                <span class="kw-2">&amp;</span>SubClones::from(
<a href=#394 id=394 data-nosnippet>394</a>                    cells_with_idx
<a href=#395 id=395 data-nosnippet>395</a>                        .into_iter()
<a href=#396 id=396 data-nosnippet>396</a>                        .map(|(cell, id)| (cell.to_owned(), id))
<a href=#397 id=397 data-nosnippet>397</a>                        .collect::&lt;Vec&lt;(StemCell, usize)&gt;&gt;(),
<a href=#398 id=398 data-nosnippet>398</a>                ),
<a href=#399 id=399 data-nosnippet>399</a>                <span class="kw-2">&amp;</span><span class="self">self</span>.make_path(Stats2Save::VariantFraction, nb_cells, time)<span class="question-mark">?</span>,
<a href=#400 id=400 data-nosnippet>400</a>                <span class="self">self</span>.verbosity,
<a href=#401 id=401 data-nosnippet>401</a>            )<span class="question-mark">?</span>;
<a href=#402 id=402 data-nosnippet>402</a>        }
<a href=#403 id=403 data-nosnippet>403</a>
<a href=#404 id=404 data-nosnippet>404</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#405 id=405 data-nosnippet>405</a>            <span class="macro">println!</span>(
<a href=#406 id=406 data-nosnippet>406</a>                <span class="string">"saved measurements after {} mutational events"</span>,
<a href=#407 id=407 data-nosnippet>407</a>                <span class="self">self</span>.counter_divisions
<a href=#408 id=408 data-nosnippet>408</a>            );
<a href=#409 id=409 data-nosnippet>409</a>        }
<a href=#410 id=410 data-nosnippet>410</a>        <span class="prelude-val">Ok</span>(())
<a href=#411 id=411 data-nosnippet>411</a>    }
<a href=#412 id=412 data-nosnippet>412</a>
<a href=#413 id=413 data-nosnippet>413</a>    <span class="kw">pub fn </span>save(
<a href=#414 id=414 data-nosnippet>414</a>        <span class="kw-2">&amp;mut </span><span class="self">self</span>,
<a href=#415 id=415 data-nosnippet>415</a>        time: f32,
<a href=#416 id=416 data-nosnippet>416</a>        saving_cells: <span class="kw-2">&amp;</span>SavingCells,
<a href=#417 id=417 data-nosnippet>417</a>        save_sfs_only: bool,
<a href=#418 id=418 data-nosnippet>418</a>        rng: <span class="kw-2">&amp;mut </span><span class="kw">impl </span>Rng,
<a href=#419 id=419 data-nosnippet>419</a>    ) -&gt; anyhow::Result&lt;()&gt; {
<a href=#420 id=420 data-nosnippet>420</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#421 id=421 data-nosnippet>421</a>            <span class="macro">println!</span>(<span class="string">"saving process at time {time}"</span>);
<a href=#422 id=422 data-nosnippet>422</a>        }
<a href=#423 id=423 data-nosnippet>423</a>        <span class="kw">if let </span>NeutralMutations::UponDivisionAndBackground = <span class="self">self</span>.proliferation.neutral_mutation {
<a href=#424 id=424 data-nosnippet>424</a>            <span class="comment">// this is important: we update all background mutations at this
<a href=#425 id=425 data-nosnippet>425</a>            // time such that all cells are all on the same page.
<a href=#426 id=426 data-nosnippet>426</a>            // Since background mutations are implemented at each division, and
<a href=#427 id=427 data-nosnippet>427</a>            // cells do not proliferate at the same rate, we need to correct and
<a href=#428 id=428 data-nosnippet>428</a>            // update the background mutations at the timepoint corresponding
<a href=#429 id=429 data-nosnippet>429</a>            // to sampling step, i.e. before saving.
<a href=#430 id=430 data-nosnippet>430</a>            </span><span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#431 id=431 data-nosnippet>431</a>                <span class="macro">println!</span>(<span class="string">"updating the neutral background mutations for all cells"</span>);
<a href=#432 id=432 data-nosnippet>432</a>            }
<a href=#433 id=433 data-nosnippet>433</a>            <span class="kw">for </span>stem_cell <span class="kw">in </span><span class="self">self</span>.subclones.get_mut_cells() {
<a href=#434 id=434 data-nosnippet>434</a>                assign_background_mutations(
<a href=#435 id=435 data-nosnippet>435</a>                    stem_cell,
<a href=#436 id=436 data-nosnippet>436</a>                    <span class="self">self</span>.time,
<a href=#437 id=437 data-nosnippet>437</a>                    <span class="kw-2">&amp;</span><span class="self">self</span>.distributions.neutral_poisson,
<a href=#438 id=438 data-nosnippet>438</a>                    rng,
<a href=#439 id=439 data-nosnippet>439</a>                    <span class="self">self</span>.verbosity,
<a href=#440 id=440 data-nosnippet>440</a>                );
<a href=#441 id=441 data-nosnippet>441</a>            }
<a href=#442 id=442 data-nosnippet>442</a>        }
<a href=#443 id=443 data-nosnippet>443</a>        <span class="kw">let </span>population = <span class="self">self</span>.subclones.get_cells_with_clones_idx();
<a href=#444 id=444 data-nosnippet>444</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#445 id=445 data-nosnippet>445</a>            <span class="macro">println!</span>(<span class="string">"saving {saving_cells:#?}"</span>);
<a href=#446 id=446 data-nosnippet>446</a>        }
<a href=#447 id=447 data-nosnippet>447</a>        <span class="kw">match </span>saving_cells {
<a href=#448 id=448 data-nosnippet>448</a>            SavingCells::WholePopulation =&gt; <span class="self">self
<a href=#449 id=449 data-nosnippet>449</a>                </span>.save_it(time, population, save_sfs_only)
<a href=#450 id=450 data-nosnippet>450</a>                .with_context(|| <span class="string">"cannot save the full population"</span>)
<a href=#451 id=451 data-nosnippet>451</a>                .unwrap(),
<a href=#452 id=452 data-nosnippet>452</a>            SavingCells::Subsampling {
<a href=#453 id=453 data-nosnippet>453</a>                sample_size,
<a href=#454 id=454 data-nosnippet>454</a>                population_as_well: populaiton_as_well,
<a href=#455 id=455 data-nosnippet>455</a>            } =&gt; {
<a href=#456 id=456 data-nosnippet>456</a>                <span class="kw">if </span><span class="kw-2">*</span>populaiton_as_well {
<a href=#457 id=457 data-nosnippet>457</a>                    <span class="self">self</span>.save_it(time, population, save_sfs_only)
<a href=#458 id=458 data-nosnippet>458</a>                        .with_context(|| <span class="string">"cannot save the full population"</span>)
<a href=#459 id=459 data-nosnippet>459</a>                        .unwrap();
<a href=#460 id=460 data-nosnippet>460</a>                }
<a href=#461 id=461 data-nosnippet>461</a>                <span class="kw">let </span>cells_with_idx = <span class="self">self
<a href=#462 id=462 data-nosnippet>462</a>                    </span>.subclones
<a href=#463 id=463 data-nosnippet>463</a>                    .get_cells_subsampled_with_clones_idx(<span class="kw-2">*</span>sample_size, rng);
<a href=#464 id=464 data-nosnippet>464</a>                <span class="macro">assert_eq!</span>(cells_with_idx.len(), <span class="kw-2">*</span>sample_size);
<a href=#465 id=465 data-nosnippet>465</a>                <span class="self">self</span>.save_it(time, cells_with_idx, save_sfs_only)
<a href=#466 id=466 data-nosnippet>466</a>                    .with_context(|| <span class="string">"cannot save the subsample"</span>)
<a href=#467 id=467 data-nosnippet>467</a>                    .unwrap();
<a href=#468 id=468 data-nosnippet>468</a>            }
<a href=#469 id=469 data-nosnippet>469</a>        }
<a href=#470 id=470 data-nosnippet>470</a>        <span class="prelude-val">Ok</span>(())
<a href=#471 id=471 data-nosnippet>471</a>    }
<a href=#472 id=472 data-nosnippet>472</a>
<a href=#473 id=473 data-nosnippet>473</a>    <span class="kw">pub fn </span>into_subsampled(<span class="self">self</span>, nb_cells: NonZeroUsize, rng: <span class="kw-2">&amp;mut </span><span class="kw">impl </span>Rng) -&gt; <span class="self">Self </span>{
<a href=#474 id=474 data-nosnippet>474</a>        <span class="doccomment">//! Take a subsample (without replacement) of the population of this
<a href=#475 id=475 data-nosnippet>475</a>        //! process using [`choose_multiple`](https://docs.rs/rand/latest/rand/seq/trait.IndexedRandom.html#method.choose_multiple).
<a href=#476 id=476 data-nosnippet>476</a>        </span><span class="kw">let </span><span class="kw-2">mut </span>moran: Moran = <span class="self">self</span>;
<a href=#477 id=477 data-nosnippet>477</a>        moran.subclones = moran.subclones.into_subsampled(nb_cells, rng);
<a href=#478 id=478 data-nosnippet>478</a>        <span class="kw">if </span>moran.verbosity &gt; <span class="number">0 </span>{
<a href=#479 id=479 data-nosnippet>479</a>            <span class="macro">println!</span>(<span class="string">"Subsampled {} cells"</span>, moran.subclones.get_cells().len())
<a href=#480 id=480 data-nosnippet>480</a>        }
<a href=#481 id=481 data-nosnippet>481</a>
<a href=#482 id=482 data-nosnippet>482</a>        moran
<a href=#483 id=483 data-nosnippet>483</a>    }
<a href=#484 id=484 data-nosnippet>484</a>}
<a href=#485 id=485 data-nosnippet>485</a>
<a href=#486 id=486 data-nosnippet>486</a><span class="kw">impl </span>AdvanceStep&lt;MAX_SUBCLONES&gt; <span class="kw">for </span>Moran {
<a href=#487 id=487 data-nosnippet>487</a>    <span class="doccomment">/// The id of a subclone [`CloneId`].
<a href=#488 id=488 data-nosnippet>488</a>    </span><span class="kw">type </span>Reaction = CloneId;
<a href=#489 id=489 data-nosnippet>489</a>
<a href=#490 id=490 data-nosnippet>490</a>    <span class="kw">fn </span>advance_step(<span class="kw-2">&amp;mut </span><span class="self">self</span>, reaction: NextReaction&lt;<span class="self">Self</span>::Reaction&gt;, rng: <span class="kw-2">&amp;mut </span><span class="kw">impl </span>Rng) {
<a href=#491 id=491 data-nosnippet>491</a>        <span class="doccomment">//! Update the process by simulating the next proliferative event
<a href=#492 id=492 data-nosnippet>492</a>        //! according to the next `reaction` determined by the [Gillespie
<a href=#493 id=493 data-nosnippet>493</a>        //! algorithm](`sosa`).
<a href=#494 id=494 data-nosnippet>494</a>        //! Here we also save the state of the simulation at different
<a href=#495 id=495 data-nosnippet>495</a>        //! timepoints, see `self.snapshot`.
<a href=#496 id=496 data-nosnippet>496</a>        //!
<a href=#497 id=497 data-nosnippet>497</a>        //! For every simulated timestep, the algorithm samples an exponential
<a href=#498 id=498 data-nosnippet>498</a>        //! waiting time and the non-empty clone that will proliferate.
<a href=#499 id=499 data-nosnippet>499</a>        //! From the proliferating clone selected, we sample a random cell to
<a href=#500 id=500 data-nosnippet>500</a>        //! which we assign neutral and fit mutations.
<a href=#501 id=501 data-nosnippet>501</a>        </span><span class="comment">// take snapshot
<a href=#502 id=502 data-nosnippet>502</a>        </span><span class="kw">while </span>!<span class="self">self</span>.snapshots.is_empty() &amp;&amp; <span class="self">self</span>.snapshots.iter().any(|s| <span class="self">self</span>.time &gt;= s.time) {
<a href=#503 id=503 data-nosnippet>503</a>            <span class="kw">let </span>snapshot = <span class="self">self</span>.snapshots.pop_front().unwrap();
<a href=#504 id=504 data-nosnippet>504</a>            <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">0 </span>{
<a href=#505 id=505 data-nosnippet>505</a>                <span class="macro">println!</span>(
<a href=#506 id=506 data-nosnippet>506</a>                    <span class="string">"saving state for timepoint at simulation's time {} for timepoint at {:#?} with {} cells"</span>,
<a href=#507 id=507 data-nosnippet>507</a>                    <span class="self">self</span>.time, snapshot.time, snapshot.cells2sample
<a href=#508 id=508 data-nosnippet>508</a>                );
<a href=#509 id=509 data-nosnippet>509</a>            }
<a href=#510 id=510 data-nosnippet>510</a>            <span class="kw">let </span>saving_cells =
<a href=#511 id=511 data-nosnippet>511</a>                <span class="kw">if </span>snapshot.cells2sample == <span class="self">self</span>.subclones.compute_tot_cells() <span class="kw">as </span>usize {
<a href=#512 id=512 data-nosnippet>512</a>                    SavingCells::WholePopulation
<a href=#513 id=513 data-nosnippet>513</a>                } <span class="kw">else </span>{
<a href=#514 id=514 data-nosnippet>514</a>                    SavingCells::Subsampling {
<a href=#515 id=515 data-nosnippet>515</a>                        sample_size: snapshot.cells2sample,
<a href=#516 id=516 data-nosnippet>516</a>                        population_as_well: <span class="self">self</span>.save_population,
<a href=#517 id=517 data-nosnippet>517</a>                    }
<a href=#518 id=518 data-nosnippet>518</a>                };
<a href=#519 id=519 data-nosnippet>519</a>            <span class="self">self</span>.save(<span class="self">self</span>.time, <span class="kw-2">&amp;</span>saving_cells, <span class="self">self</span>.save_sfs_only, rng)
<a href=#520 id=520 data-nosnippet>520</a>                .expect(<span class="string">"cannot save snapshot"</span>);
<a href=#521 id=521 data-nosnippet>521</a>        }
<a href=#522 id=522 data-nosnippet>522</a>
<a href=#523 id=523 data-nosnippet>523</a>        <span class="comment">// 1. select a cell `c` at random
<a href=#524 id=524 data-nosnippet>524</a>        // the SSA samples the clone that will proliferate next, that is the
<a href=#525 id=525 data-nosnippet>525</a>        // clone with id `reaction.event`. Pick random proliferating cells from
<a href=#526 id=526 data-nosnippet>526</a>        // this clone. **Note that this removes the cells from the clone with
<a href=#527 id=527 data-nosnippet>527</a>        // id `reaction.event`.**
<a href=#528 id=528 data-nosnippet>528</a>        </span><span class="self">self</span>.time += reaction.time;
<a href=#529 id=529 data-nosnippet>529</a>        <span class="self">self</span>.counter_divisions += <span class="number">1</span>;
<a href=#530 id=530 data-nosnippet>530</a>        <span class="self">self</span>.proliferation.proliferate(
<a href=#531 id=531 data-nosnippet>531</a>            <span class="kw-2">&amp;mut </span><span class="self">self</span>.subclones,
<a href=#532 id=532 data-nosnippet>532</a>            <span class="self">self</span>.time,
<a href=#533 id=533 data-nosnippet>533</a>            reaction.event,
<a href=#534 id=534 data-nosnippet>534</a>            <span class="kw-2">&amp;</span><span class="self">self</span>.distributions,
<a href=#535 id=535 data-nosnippet>535</a>            rng,
<a href=#536 id=536 data-nosnippet>536</a>            <span class="self">self</span>.verbosity,
<a href=#537 id=537 data-nosnippet>537</a>        );
<a href=#538 id=538 data-nosnippet>538</a>
<a href=#539 id=539 data-nosnippet>539</a>        <span class="comment">// remove a cell from the population
<a href=#540 id=540 data-nosnippet>540</a>        </span><span class="self">self</span>.keep_const_population_upon_symmetric_division(rng);
<a href=#541 id=541 data-nosnippet>541</a>
<a href=#542 id=542 data-nosnippet>542</a>        <span class="kw">if </span><span class="self">self</span>.verbosity &gt; <span class="number">2 </span>{
<a href=#543 id=543 data-nosnippet>543</a>            <span class="macro">println!</span>(<span class="string">"{} cells"</span>, <span class="self">self</span>.subclones.compute_tot_cells());
<a href=#544 id=544 data-nosnippet>544</a>        }
<a href=#545 id=545 data-nosnippet>545</a>    }
<a href=#546 id=546 data-nosnippet>546</a>
<a href=#547 id=547 data-nosnippet>547</a>    <span class="kw">fn </span>update_state(<span class="kw-2">&amp;</span><span class="self">self</span>, state: <span class="kw-2">&amp;mut </span>CurrentState&lt;MAX_SUBCLONES&gt;) {
<a href=#548 id=548 data-nosnippet>548</a>        state.population = Variants::variant_counts(<span class="kw-2">&amp;</span><span class="self">self</span>.subclones);
<a href=#549 id=549 data-nosnippet>549</a>    }
<a href=#550 id=550 data-nosnippet>550</a>}
<a href=#551 id=551 data-nosnippet>551</a>
<a href=#552 id=552 data-nosnippet>552</a><span class="attr">#[cfg(test)]
<a href=#553 id=553 data-nosnippet>553</a></span><span class="kw">mod </span>tests {
<a href=#554 id=554 data-nosnippet>554</a>    <span class="kw">use </span>std::num::{NonZeroU64, NonZeroU8};
<a href=#555 id=555 data-nosnippet>555</a>
<a href=#556 id=556 data-nosnippet>556</a>    <span class="kw">use </span>quickcheck_macros::quickcheck;
<a href=#557 id=557 data-nosnippet>557</a>    <span class="kw">use </span>rand::SeedableRng;
<a href=#558 id=558 data-nosnippet>558</a>    <span class="kw">use </span>rand_chacha::ChaCha8Rng;
<a href=#559 id=559 data-nosnippet>559</a>
<a href=#560 id=560 data-nosnippet>560</a>    <span class="kw">use crate</span>::{genotype, Probs};
<a href=#561 id=561 data-nosnippet>561</a>
<a href=#562 id=562 data-nosnippet>562</a>    <span class="kw">use super</span>::<span class="kw-2">*</span>;
<a href=#563 id=563 data-nosnippet>563</a>
<a href=#564 id=564 data-nosnippet>564</a>    <span class="kw">struct </span>MoranFitVariant {
<a href=#565 id=565 data-nosnippet>565</a>        tot_cells: u64,
<a href=#566 id=566 data-nosnippet>566</a>        process: Moran,
<a href=#567 id=567 data-nosnippet>567</a>    }
<a href=#568 id=568 data-nosnippet>568</a>
<a href=#569 id=569 data-nosnippet>569</a>    <span class="kw">struct </span>MoranNoFitVariant {
<a href=#570 id=570 data-nosnippet>570</a>        tot_cells: u64,
<a href=#571 id=571 data-nosnippet>571</a>        process: Moran,
<a href=#572 id=572 data-nosnippet>572</a>    }
<a href=#573 id=573 data-nosnippet>573</a>
<a href=#574 id=574 data-nosnippet>574</a>    <span class="kw">fn </span>create_moran_fit_variants(cells: NonZeroU64) -&gt; MoranFitVariant {
<a href=#575 id=575 data-nosnippet>575</a>        MoranFitVariant {
<a href=#576 id=576 data-nosnippet>576</a>            tot_cells: cells.get(),
<a href=#577 id=577 data-nosnippet>577</a>            process: create_moran(<span class="bool-val">true</span>, cells),
<a href=#578 id=578 data-nosnippet>578</a>        }
<a href=#579 id=579 data-nosnippet>579</a>    }
<a href=#580 id=580 data-nosnippet>580</a>
<a href=#581 id=581 data-nosnippet>581</a>    <span class="kw">fn </span>create_moran_no_fit_variants(cells: NonZeroU64) -&gt; MoranNoFitVariant {
<a href=#582 id=582 data-nosnippet>582</a>        MoranNoFitVariant {
<a href=#583 id=583 data-nosnippet>583</a>            tot_cells: cells.get(),
<a href=#584 id=584 data-nosnippet>584</a>            process: create_moran(<span class="bool-val">false</span>, cells),
<a href=#585 id=585 data-nosnippet>585</a>        }
<a href=#586 id=586 data-nosnippet>586</a>    }
<a href=#587 id=587 data-nosnippet>587</a>
<a href=#588 id=588 data-nosnippet>588</a>    <span class="kw">fn </span>create_moran(fit_variants: bool, cells: NonZeroU64) -&gt; Moran {
<a href=#589 id=589 data-nosnippet>589</a>        <span class="kw">let </span>mu = <span class="kw">if </span>fit_variants { <span class="number">0.999 </span>} <span class="kw">else </span>{ <span class="number">0. </span>};
<a href=#590 id=590 data-nosnippet>590</a>        Moran::new(
<a href=#591 id=591 data-nosnippet>591</a>            ProcessOptions {
<a href=#592 id=592 data-nosnippet>592</a>                path: PathBuf::default(),
<a href=#593 id=593 data-nosnippet>593</a>                snapshots: VecDeque::default(),
<a href=#594 id=594 data-nosnippet>594</a>            },
<a href=#595 id=595 data-nosnippet>595</a>            SubClones::new(<span class="macro">vec!</span>[StemCell::new(); cells.get() <span class="kw">as </span>usize], <span class="number">3</span>, <span class="number">0</span>),
<a href=#596 id=596 data-nosnippet>596</a>            <span class="number">0f32</span>,
<a href=#597 id=597 data-nosnippet>597</a>            SavingOptions {
<a href=#598 id=598 data-nosnippet>598</a>                filename: PathBuf::default(),
<a href=#599 id=599 data-nosnippet>599</a>                save_sfs_only: <span class="bool-val">true</span>,
<a href=#600 id=600 data-nosnippet>600</a>                save_population: <span class="bool-val">true</span>,
<a href=#601 id=601 data-nosnippet>601</a>            },
<a href=#602 id=602 data-nosnippet>602</a>            Distributions::new(Probs::new(<span class="number">1.</span>, <span class="number">1.</span>, mu, <span class="number">1.</span>, cells.get(), <span class="number">0</span>), <span class="number">0</span>),
<a href=#603 id=603 data-nosnippet>603</a>            Proliferation::default(),
<a href=#604 id=604 data-nosnippet>604</a>            <span class="number">0</span>,
<a href=#605 id=605 data-nosnippet>605</a>        )
<a href=#606 id=606 data-nosnippet>606</a>    }
<a href=#607 id=607 data-nosnippet>607</a>
<a href=#608 id=608 data-nosnippet>608</a>    <span class="kw">fn </span>assert_pop_const(moran: <span class="kw-2">&amp;</span>Moran, cells: u64) {
<a href=#609 id=609 data-nosnippet>609</a>        <span class="macro">assert_eq!</span>(moran.subclones.compute_tot_cells(), cells);
<a href=#610 id=610 data-nosnippet>610</a>    }
<a href=#611 id=611 data-nosnippet>611</a>
<a href=#612 id=612 data-nosnippet>612</a>    <span class="kw">fn </span>assert_not_all_cells_in_wild_type(moran: <span class="kw-2">&amp;</span>Moran, cells: u64) {
<a href=#613 id=613 data-nosnippet>613</a>        <span class="macro">assert_ne!</span>(moran.subclones.get_clone(<span class="number">0</span>).unwrap().cell_count(), cells);
<a href=#614 id=614 data-nosnippet>614</a>    }
<a href=#615 id=615 data-nosnippet>615</a>
<a href=#616 id=616 data-nosnippet>616</a>    <span class="kw">fn </span>assert_all_cells_in_wild_type(moran: <span class="kw-2">&amp;</span>Moran, cells: u64) {
<a href=#617 id=617 data-nosnippet>617</a>        <span class="macro">assert_eq!</span>(moran.subclones.get_clone(<span class="number">0</span>).unwrap().cell_count(), cells);
<a href=#618 id=618 data-nosnippet>618</a>        <span class="macro">assert_eq!</span>(the_only_one_subclone_present(<span class="kw-2">&amp;</span>moran.subclones).unwrap(), <span class="number">0</span>);
<a href=#619 id=619 data-nosnippet>619</a>    }
<a href=#620 id=620 data-nosnippet>620</a>
<a href=#621 id=621 data-nosnippet>621</a>    <span class="attr">#[quickcheck]
<a href=#622 id=622 data-nosnippet>622</a>    </span><span class="kw">fn </span>advance_moran_no_fit_variant_test(cells: NonZeroU8, seed: u64) {
<a href=#623 id=623 data-nosnippet>623</a>        <span class="kw">let </span><span class="kw-2">mut </span>moran = create_moran_no_fit_variants(NonZeroU64::new(cells.get() <span class="kw">as </span>u64).unwrap());
<a href=#624 id=624 data-nosnippet>624</a>        <span class="kw">let </span>reaction = NextReaction {
<a href=#625 id=625 data-nosnippet>625</a>            time: <span class="number">12.2</span>,
<a href=#626 id=626 data-nosnippet>626</a>            event: <span class="number">0</span>,
<a href=#627 id=627 data-nosnippet>627</a>        };
<a href=#628 id=628 data-nosnippet>628</a>        <span class="kw">let </span>burden_before = genotype::MutationalBurden::from_moran(<span class="kw-2">&amp;</span>moran.process, <span class="number">0</span>).unwrap();
<a href=#629 id=629 data-nosnippet>629</a>        assert_pop_const(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#630 id=630 data-nosnippet>630</a>        assert_all_cells_in_wild_type(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#631 id=631 data-nosnippet>631</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#632 id=632 data-nosnippet>632</a>
<a href=#633 id=633 data-nosnippet>633</a>        moran.process.advance_step(reaction, rng);
<a href=#634 id=634 data-nosnippet>634</a>
<a href=#635 id=635 data-nosnippet>635</a>        assert_pop_const(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#636 id=636 data-nosnippet>636</a>        assert_all_cells_in_wild_type(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#637 id=637 data-nosnippet>637</a>        <span class="kw">let </span>burden_after = genotype::MutationalBurden::from_moran(<span class="kw-2">&amp;</span>moran.process, <span class="number">0</span>).unwrap();
<a href=#638 id=638 data-nosnippet>638</a>        <span class="macro">assert!</span>(burden_before.<span class="number">0</span>.keys().sum::&lt;u16&gt;() &lt; burden_after.<span class="number">0</span>.keys().sum::&lt;u16&gt;());
<a href=#639 id=639 data-nosnippet>639</a>    }
<a href=#640 id=640 data-nosnippet>640</a>
<a href=#641 id=641 data-nosnippet>641</a>    <span class="attr">#[quickcheck]
<a href=#642 id=642 data-nosnippet>642</a>    </span><span class="kw">fn </span>advance_moran_test(cells: NonZeroU8, seed: u64, new_clone: bool) {
<a href=#643 id=643 data-nosnippet>643</a>        <span class="kw">let </span><span class="kw-2">mut </span>moran = create_moran_fit_variants(NonZeroU64::new(cells.get() <span class="kw">as </span>u64).unwrap());
<a href=#644 id=644 data-nosnippet>644</a>        <span class="kw">let </span>reaction = <span class="kw">if </span>new_clone {
<a href=#645 id=645 data-nosnippet>645</a>            NextReaction {
<a href=#646 id=646 data-nosnippet>646</a>                time: cells.get() <span class="kw">as </span>f32 * <span class="number">0.9999</span>,
<a href=#647 id=647 data-nosnippet>647</a>                event: <span class="number">0</span>,
<a href=#648 id=648 data-nosnippet>648</a>            }
<a href=#649 id=649 data-nosnippet>649</a>        } <span class="kw">else </span>{
<a href=#650 id=650 data-nosnippet>650</a>            NextReaction {
<a href=#651 id=651 data-nosnippet>651</a>                time: <span class="number">0.0002</span>,
<a href=#652 id=652 data-nosnippet>652</a>                event: <span class="number">0</span>,
<a href=#653 id=653 data-nosnippet>653</a>            }
<a href=#654 id=654 data-nosnippet>654</a>        };
<a href=#655 id=655 data-nosnippet>655</a>        <span class="kw">let </span>burden_before = genotype::MutationalBurden::from_moran(<span class="kw-2">&amp;</span>moran.process, <span class="number">0</span>).unwrap();
<a href=#656 id=656 data-nosnippet>656</a>        assert_pop_const(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#657 id=657 data-nosnippet>657</a>        assert_all_cells_in_wild_type(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#658 id=658 data-nosnippet>658</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#659 id=659 data-nosnippet>659</a>
<a href=#660 id=660 data-nosnippet>660</a>        moran.process.advance_step(reaction, rng);
<a href=#661 id=661 data-nosnippet>661</a>
<a href=#662 id=662 data-nosnippet>662</a>        assert_pop_const(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#663 id=663 data-nosnippet>663</a>        <span class="kw">if </span>new_clone {
<a href=#664 id=664 data-nosnippet>664</a>            assert_not_all_cells_in_wild_type(<span class="kw-2">&amp;</span>moran.process, moran.tot_cells);
<a href=#665 id=665 data-nosnippet>665</a>        }
<a href=#666 id=666 data-nosnippet>666</a>        <span class="kw">let </span>burden_after = genotype::MutationalBurden::from_moran(<span class="kw-2">&amp;</span>moran.process, <span class="number">0</span>).unwrap();
<a href=#667 id=667 data-nosnippet>667</a>        <span class="macro">assert!</span>(burden_before.<span class="number">0</span>.keys().sum::&lt;u16&gt;() &lt;= burden_after.<span class="number">0</span>.keys().sum::&lt;u16&gt;());
<a href=#668 id=668 data-nosnippet>668</a>    }
<a href=#669 id=669 data-nosnippet>669</a>
<a href=#670 id=670 data-nosnippet>670</a>    <span class="kw">struct </span>ExpFitVariant {
<a href=#671 id=671 data-nosnippet>671</a>        tot_cells: u64,
<a href=#672 id=672 data-nosnippet>672</a>        process: Exponential,
<a href=#673 id=673 data-nosnippet>673</a>    }
<a href=#674 id=674 data-nosnippet>674</a>
<a href=#675 id=675 data-nosnippet>675</a>    <span class="kw">struct </span>ExpNoFitVariant {
<a href=#676 id=676 data-nosnippet>676</a>        tot_cells: u64,
<a href=#677 id=677 data-nosnippet>677</a>        process: Exponential,
<a href=#678 id=678 data-nosnippet>678</a>    }
<a href=#679 id=679 data-nosnippet>679</a>
<a href=#680 id=680 data-nosnippet>680</a>    <span class="kw">fn </span>create_exp_fit_variants(cells: NonZeroU64, is_regrowth: bool) -&gt; ExpFitVariant {
<a href=#681 id=681 data-nosnippet>681</a>        ExpFitVariant {
<a href=#682 id=682 data-nosnippet>682</a>            tot_cells: cells.get(),
<a href=#683 id=683 data-nosnippet>683</a>            process: create_exp(<span class="bool-val">true</span>, cells, is_regrowth),
<a href=#684 id=684 data-nosnippet>684</a>        }
<a href=#685 id=685 data-nosnippet>685</a>    }
<a href=#686 id=686 data-nosnippet>686</a>
<a href=#687 id=687 data-nosnippet>687</a>    <span class="kw">fn </span>create_exp_no_fit_variants(cells: NonZeroU64, is_regrowth: bool) -&gt; ExpNoFitVariant {
<a href=#688 id=688 data-nosnippet>688</a>        ExpNoFitVariant {
<a href=#689 id=689 data-nosnippet>689</a>            tot_cells: cells.get(),
<a href=#690 id=690 data-nosnippet>690</a>            process: create_exp(<span class="bool-val">false</span>, cells, is_regrowth),
<a href=#691 id=691 data-nosnippet>691</a>        }
<a href=#692 id=692 data-nosnippet>692</a>    }
<a href=#693 id=693 data-nosnippet>693</a>
<a href=#694 id=694 data-nosnippet>694</a>    <span class="kw">fn </span>create_exp(fit_variants: bool, cells: NonZeroU64, is_regrowht: bool) -&gt; Exponential {
<a href=#695 id=695 data-nosnippet>695</a>        <span class="kw">let </span>mu = <span class="kw">if </span>fit_variants { <span class="number">0.999 </span>} <span class="kw">else </span>{ <span class="number">0. </span>};
<a href=#696 id=696 data-nosnippet>696</a>        <span class="kw">let </span>phase = <span class="kw">if </span>is_regrowht {
<a href=#697 id=697 data-nosnippet>697</a>            ExponentialPhase::Regrowth
<a href=#698 id=698 data-nosnippet>698</a>        } <span class="kw">else </span>{
<a href=#699 id=699 data-nosnippet>699</a>            ExponentialPhase::Development
<a href=#700 id=700 data-nosnippet>700</a>        };
<a href=#701 id=701 data-nosnippet>701</a>        Exponential::new(
<a href=#702 id=702 data-nosnippet>702</a>            SubClones::new(<span class="macro">vec!</span>[StemCell::new(); cells.get() <span class="kw">as </span>usize], <span class="number">3</span>, <span class="number">0</span>),
<a href=#703 id=703 data-nosnippet>703</a>            Distributions::new(Probs::new(<span class="number">10.</span>, <span class="number">1.</span>, mu, <span class="number">1.</span>, cells.get(), <span class="number">0</span>), <span class="number">0</span>),
<a href=#704 id=704 data-nosnippet>704</a>            Proliferation::default(),
<a href=#705 id=705 data-nosnippet>705</a>            <span class="number">0.</span>,
<a href=#706 id=706 data-nosnippet>706</a>            phase,
<a href=#707 id=707 data-nosnippet>707</a>            <span class="number">0</span>,
<a href=#708 id=708 data-nosnippet>708</a>        )
<a href=#709 id=709 data-nosnippet>709</a>    }
<a href=#710 id=710 data-nosnippet>710</a>
<a href=#711 id=711 data-nosnippet>711</a>    <span class="kw">fn </span>assert_not_all_cells_in_wild_type_exp(exp: <span class="kw-2">&amp;</span>Exponential, cells: u64) {
<a href=#712 id=712 data-nosnippet>712</a>        <span class="macro">assert_ne!</span>(exp.subclones.get_clone(<span class="number">0</span>).unwrap().cell_count(), cells);
<a href=#713 id=713 data-nosnippet>713</a>    }
<a href=#714 id=714 data-nosnippet>714</a>
<a href=#715 id=715 data-nosnippet>715</a>    <span class="kw">fn </span>assert_all_cells_in_wild_type_exp(exp: <span class="kw-2">&amp;</span>Exponential, cells: u64) {
<a href=#716 id=716 data-nosnippet>716</a>        <span class="macro">assert_eq!</span>(exp.subclones.get_clone(<span class="number">0</span>).unwrap().cell_count(), cells);
<a href=#717 id=717 data-nosnippet>717</a>        <span class="macro">assert_eq!</span>(the_only_one_subclone_present(<span class="kw-2">&amp;</span>exp.subclones).unwrap(), <span class="number">0</span>);
<a href=#718 id=718 data-nosnippet>718</a>    }
<a href=#719 id=719 data-nosnippet>719</a>
<a href=#720 id=720 data-nosnippet>720</a>    <span class="attr">#[quickcheck]
<a href=#721 id=721 data-nosnippet>721</a>    </span><span class="kw">fn </span>advance_exp_no_fit_variant_test(cells: NonZeroU8, seed: u64) {
<a href=#722 id=722 data-nosnippet>722</a>        <span class="kw">let </span><span class="kw-2">mut </span>exp =
<a href=#723 id=723 data-nosnippet>723</a>            create_exp_no_fit_variants(NonZeroU64::new(cells.get() <span class="kw">as </span>u64).unwrap(), <span class="bool-val">false</span>);
<a href=#724 id=724 data-nosnippet>724</a>        <span class="kw">let </span>reaction = NextReaction {
<a href=#725 id=725 data-nosnippet>725</a>            time: <span class="number">12.2</span>,
<a href=#726 id=726 data-nosnippet>726</a>            event: <span class="number">0</span>,
<a href=#727 id=727 data-nosnippet>727</a>        };
<a href=#728 id=728 data-nosnippet>728</a>        <span class="kw">let </span>burden_before = genotype::MutationalBurden::from_exp(<span class="kw-2">&amp;</span>exp.process, <span class="number">0</span>).unwrap();
<a href=#729 id=729 data-nosnippet>729</a>        <span class="kw">let </span>old_cells = exp.process.subclones.compute_tot_cells();
<a href=#730 id=730 data-nosnippet>730</a>        assert_all_cells_in_wild_type_exp(<span class="kw-2">&amp;</span>exp.process, exp.tot_cells);
<a href=#731 id=731 data-nosnippet>731</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#732 id=732 data-nosnippet>732</a>
<a href=#733 id=733 data-nosnippet>733</a>        exp.process.advance_step(reaction, rng);
<a href=#734 id=734 data-nosnippet>734</a>
<a href=#735 id=735 data-nosnippet>735</a>        <span class="macro">assert_eq!</span>(exp.process.subclones.compute_tot_cells(), old_cells + <span class="number">1</span>);
<a href=#736 id=736 data-nosnippet>736</a>        <span class="kw">let </span>burden_after = genotype::MutationalBurden::from_exp(<span class="kw-2">&amp;</span>exp.process, <span class="number">0</span>).unwrap();
<a href=#737 id=737 data-nosnippet>737</a>        <span class="macro">assert!</span>(burden_before.<span class="number">0</span>.keys().sum::&lt;u16&gt;() &lt;= burden_after.<span class="number">0</span>.keys().sum::&lt;u16&gt;());
<a href=#738 id=738 data-nosnippet>738</a>    }
<a href=#739 id=739 data-nosnippet>739</a>
<a href=#740 id=740 data-nosnippet>740</a>    <span class="attr">#[quickcheck]
<a href=#741 id=741 data-nosnippet>741</a>    </span><span class="kw">fn </span>advance_exp_test(cells: NonZeroU8, seed: u64, new_clone: bool) {
<a href=#742 id=742 data-nosnippet>742</a>        <span class="kw">let </span><span class="kw-2">mut </span>exp = create_exp_fit_variants(NonZeroU64::new(cells.get() <span class="kw">as </span>u64).unwrap(), <span class="bool-val">false</span>);
<a href=#743 id=743 data-nosnippet>743</a>        <span class="kw">let </span>reaction = <span class="kw">if </span>new_clone {
<a href=#744 id=744 data-nosnippet>744</a>            NextReaction {
<a href=#745 id=745 data-nosnippet>745</a>                time: cells.get() <span class="kw">as </span>f32 * <span class="number">0.9999</span>,
<a href=#746 id=746 data-nosnippet>746</a>                event: <span class="number">0</span>,
<a href=#747 id=747 data-nosnippet>747</a>            }
<a href=#748 id=748 data-nosnippet>748</a>        } <span class="kw">else </span>{
<a href=#749 id=749 data-nosnippet>749</a>            NextReaction {
<a href=#750 id=750 data-nosnippet>750</a>                time: <span class="number">0.0002</span>,
<a href=#751 id=751 data-nosnippet>751</a>                event: <span class="number">0</span>,
<a href=#752 id=752 data-nosnippet>752</a>            }
<a href=#753 id=753 data-nosnippet>753</a>        };
<a href=#754 id=754 data-nosnippet>754</a>        <span class="kw">let </span>burden_before = genotype::MutationalBurden::from_exp(<span class="kw-2">&amp;</span>exp.process, <span class="number">0</span>).unwrap();
<a href=#755 id=755 data-nosnippet>755</a>        <span class="kw">let </span>old_cells = exp.process.subclones.compute_tot_cells();
<a href=#756 id=756 data-nosnippet>756</a>        assert_all_cells_in_wild_type_exp(<span class="kw-2">&amp;</span>exp.process, exp.tot_cells);
<a href=#757 id=757 data-nosnippet>757</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#758 id=758 data-nosnippet>758</a>
<a href=#759 id=759 data-nosnippet>759</a>        exp.process.advance_step(reaction, rng);
<a href=#760 id=760 data-nosnippet>760</a>
<a href=#761 id=761 data-nosnippet>761</a>        <span class="macro">assert_eq!</span>(exp.process.subclones.compute_tot_cells(), old_cells + <span class="number">1</span>);
<a href=#762 id=762 data-nosnippet>762</a>        <span class="kw">if </span>new_clone {
<a href=#763 id=763 data-nosnippet>763</a>            assert_not_all_cells_in_wild_type_exp(<span class="kw-2">&amp;</span>exp.process, exp.tot_cells);
<a href=#764 id=764 data-nosnippet>764</a>        }
<a href=#765 id=765 data-nosnippet>765</a>        <span class="kw">let </span>burden_after = genotype::MutationalBurden::from_exp(<span class="kw-2">&amp;</span>exp.process, <span class="number">0</span>).unwrap();
<a href=#766 id=766 data-nosnippet>766</a>        <span class="macro">assert!</span>(burden_before.<span class="number">0</span>.keys().sum::&lt;u16&gt;() &lt;= burden_after.<span class="number">0</span>.keys().sum::&lt;u16&gt;());
<a href=#767 id=767 data-nosnippet>767</a>    }
<a href=#768 id=768 data-nosnippet>768</a>
<a href=#769 id=769 data-nosnippet>769</a>    <span class="attr">#[quickcheck]
<a href=#770 id=770 data-nosnippet>770</a>    </span><span class="kw">fn </span>switch_to_moran_development_test(cells: NonZeroU8, seed: u64) {
<a href=#771 id=771 data-nosnippet>771</a>        <span class="kw">let </span>exp = create_exp_fit_variants(NonZeroU64::new(cells.get() <span class="kw">as </span>u64).unwrap(), <span class="bool-val">false</span>);
<a href=#772 id=772 data-nosnippet>772</a>        <span class="kw">let </span>burden_before = genotype::MutationalBurden::from_exp(<span class="kw-2">&amp;</span>exp.process, <span class="number">0</span>).unwrap();
<a href=#773 id=773 data-nosnippet>773</a>        <span class="kw">let </span>old_cells = exp.process.subclones.compute_tot_cells();
<a href=#774 id=774 data-nosnippet>774</a>        assert_all_cells_in_wild_type_exp(<span class="kw-2">&amp;</span>exp.process, exp.tot_cells);
<a href=#775 id=775 data-nosnippet>775</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#776 id=776 data-nosnippet>776</a>        <span class="kw">let </span>moran = switch_to_moran(
<a href=#777 id=777 data-nosnippet>777</a>            exp.process,
<a href=#778 id=778 data-nosnippet>778</a>            ProcessOptions::default(),
<a href=#779 id=779 data-nosnippet>779</a>            Distributions::default(),
<a href=#780 id=780 data-nosnippet>780</a>            PathBuf::default(),
<a href=#781 id=781 data-nosnippet>781</a>            <span class="bool-val">false</span>,
<a href=#782 id=782 data-nosnippet>782</a>            <span class="bool-val">false</span>,
<a href=#783 id=783 data-nosnippet>783</a>            rng,
<a href=#784 id=784 data-nosnippet>784</a>        );
<a href=#785 id=785 data-nosnippet>785</a>
<a href=#786 id=786 data-nosnippet>786</a>        <span class="kw">let </span>burden_after = genotype::MutationalBurden::from_moran(<span class="kw-2">&amp;</span>moran, <span class="number">0</span>).unwrap();
<a href=#787 id=787 data-nosnippet>787</a>        <span class="macro">assert!</span>(burden_before.<span class="number">0</span>.keys().sum::&lt;u16&gt;() &lt;= burden_after.<span class="number">0</span>.keys().sum::&lt;u16&gt;());
<a href=#788 id=788 data-nosnippet>788</a>        <span class="macro">assert_eq!</span>(old_cells, moran.subclones.compute_tot_cells());
<a href=#789 id=789 data-nosnippet>789</a>    }
<a href=#790 id=790 data-nosnippet>790</a>
<a href=#791 id=791 data-nosnippet>791</a>    <span class="attr">#[quickcheck]
<a href=#792 id=792 data-nosnippet>792</a>    </span><span class="kw">fn </span>switch_to_moran_regrowth_test(cells: NonZeroU8, seed: u64) {
<a href=#793 id=793 data-nosnippet>793</a>        <span class="kw">let </span>exp = create_exp_fit_variants(NonZeroU64::new(cells.get() <span class="kw">as </span>u64).unwrap(), <span class="bool-val">true</span>);
<a href=#794 id=794 data-nosnippet>794</a>        <span class="kw">let </span>burden_before = genotype::MutationalBurden::from_exp(<span class="kw-2">&amp;</span>exp.process, <span class="number">0</span>).unwrap();
<a href=#795 id=795 data-nosnippet>795</a>        <span class="kw">let </span>old_cells = exp.process.subclones.compute_tot_cells();
<a href=#796 id=796 data-nosnippet>796</a>        assert_all_cells_in_wild_type_exp(<span class="kw-2">&amp;</span>exp.process, exp.tot_cells);
<a href=#797 id=797 data-nosnippet>797</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#798 id=798 data-nosnippet>798</a>        <span class="kw">let </span>moran = switch_to_moran(
<a href=#799 id=799 data-nosnippet>799</a>            exp.process,
<a href=#800 id=800 data-nosnippet>800</a>            ProcessOptions::default(),
<a href=#801 id=801 data-nosnippet>801</a>            Distributions::default(),
<a href=#802 id=802 data-nosnippet>802</a>            PathBuf::default(),
<a href=#803 id=803 data-nosnippet>803</a>            <span class="bool-val">false</span>,
<a href=#804 id=804 data-nosnippet>804</a>            <span class="bool-val">false</span>,
<a href=#805 id=805 data-nosnippet>805</a>            rng,
<a href=#806 id=806 data-nosnippet>806</a>        );
<a href=#807 id=807 data-nosnippet>807</a>
<a href=#808 id=808 data-nosnippet>808</a>        <span class="kw">let </span>burden_after = genotype::MutationalBurden::from_moran(<span class="kw-2">&amp;</span>moran, <span class="number">0</span>).unwrap();
<a href=#809 id=809 data-nosnippet>809</a>        <span class="macro">assert_eq!</span>(
<a href=#810 id=810 data-nosnippet>810</a>            burden_before.<span class="number">0</span>.keys().sum::&lt;u16&gt;(),
<a href=#811 id=811 data-nosnippet>811</a>            burden_after.<span class="number">0</span>.keys().sum::&lt;u16&gt;()
<a href=#812 id=812 data-nosnippet>812</a>        );
<a href=#813 id=813 data-nosnippet>813</a>        <span class="macro">assert_eq!</span>(old_cells, moran.subclones.compute_tot_cells());
<a href=#814 id=814 data-nosnippet>814</a>    }
<a href=#815 id=815 data-nosnippet>815</a>
<a href=#816 id=816 data-nosnippet>816</a>    <span class="kw">fn </span>the_only_one_subclone_present(subclones: <span class="kw-2">&amp;</span>SubClones) -&gt; <span class="prelude-ty">Option</span>&lt;CloneId&gt; {
<a href=#817 id=817 data-nosnippet>817</a>        <span class="doccomment">//! returns `None` if more than one subclone is present else the clone
<a href=#818 id=818 data-nosnippet>818</a>        //! id.
<a href=#819 id=819 data-nosnippet>819</a>        //!
<a href=#820 id=820 data-nosnippet>820</a>        //! **warning** this is very slow.
<a href=#821 id=821 data-nosnippet>821</a>        </span><span class="kw">let </span>tot_cells = subclones.compute_tot_cells();
<a href=#822 id=822 data-nosnippet>822</a>        (<span class="number">0</span>..MAX_SUBCLONES).find(|<span class="kw-2">&amp;</span>id| subclones.get_clone(id).unwrap().cell_count() == tot_cells)
<a href=#823 id=823 data-nosnippet>823</a>    }
<a href=#824 id=824 data-nosnippet>824</a>
<a href=#825 id=825 data-nosnippet>825</a>    <span class="attr">#[quickcheck]
<a href=#826 id=826 data-nosnippet>826</a>    </span><span class="kw">fn </span>subsample_moran_to_same_size(seed: u64) {
<a href=#827 id=827 data-nosnippet>827</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#828 id=828 data-nosnippet>828</a>        <span class="kw">let </span>moran = Moran::default();
<a href=#829 id=829 data-nosnippet>829</a>        <span class="kw">let </span>nb_cells_before = moran.subclones.get_cells().len();
<a href=#830 id=830 data-nosnippet>830</a>        <span class="kw">let </span>cells = nb_cells_before;
<a href=#831 id=831 data-nosnippet>831</a>        <span class="kw">let </span>subsampled = moran.into_subsampled(NonZeroUsize::new(cells).unwrap(), rng);
<a href=#832 id=832 data-nosnippet>832</a>        <span class="macro">assert!</span>(the_only_one_subclone_present(<span class="kw-2">&amp;</span>subsampled.subclones).is_none());
<a href=#833 id=833 data-nosnippet>833</a>        <span class="macro">assert!</span>(nb_cells_before == subsampled.subclones.get_cells().len());
<a href=#834 id=834 data-nosnippet>834</a>    }
<a href=#835 id=835 data-nosnippet>835</a>
<a href=#836 id=836 data-nosnippet>836</a>    <span class="attr">#[quickcheck]
<a href=#837 id=837 data-nosnippet>837</a>    </span><span class="kw">fn </span>subsample_moran(cells: NonZeroU8, seed: u64) {
<a href=#838 id=838 data-nosnippet>838</a>        <span class="kw">let </span>rng = <span class="kw-2">&amp;mut </span>ChaCha8Rng::seed_from_u64(seed);
<a href=#839 id=839 data-nosnippet>839</a>        <span class="kw">let </span>moran = Moran::default();
<a href=#840 id=840 data-nosnippet>840</a>        <span class="kw">let </span>nb_cells_before = moran.subclones.get_cells().len();
<a href=#841 id=841 data-nosnippet>841</a>        <span class="kw">let </span>subsampled =
<a href=#842 id=842 data-nosnippet>842</a>            moran.into_subsampled(NonZeroUsize::new(cells.get() <span class="kw">as </span>usize).unwrap(), rng);
<a href=#843 id=843 data-nosnippet>843</a>        <span class="kw">if </span>cells.get() == <span class="number">1 </span>{
<a href=#844 id=844 data-nosnippet>844</a>            <span class="macro">assert!</span>(the_only_one_subclone_present(<span class="kw-2">&amp;</span>subsampled.subclones).is_some());
<a href=#845 id=845 data-nosnippet>845</a>        } <span class="kw">else </span>{
<a href=#846 id=846 data-nosnippet>846</a>            <span class="macro">assert!</span>(the_only_one_subclone_present(<span class="kw-2">&amp;</span>subsampled.subclones).is_none());
<a href=#847 id=847 data-nosnippet>847</a>        }
<a href=#848 id=848 data-nosnippet>848</a>        <span class="comment">// this is always true as here we subsample to max 255 cells
<a href=#849 id=849 data-nosnippet>849</a>        </span><span class="macro">assert!</span>(nb_cells_before &gt; subsampled.subclones.get_cells().len());
<a href=#850 id=850 data-nosnippet>850</a>    }
<a href=#851 id=851 data-nosnippet>851</a>}</code></pre></div></section></main></body></html>